showing changes

mg979 opened the pull request:
Don't accept this PR. It's here to make a diff so I can show the the changes

mg979 committed:
showing changes

mg979 commented on code_map.py:
```diff
@@ -226,7 +229,7 @@ def generate_from(file):
 
-    if not map_view:
+    if not map_view or view.settings().get('is_widget'):
         return
```
this is important or it would try to map widgets

mg979 commented on code_map.py:
```diff
@@ -492,16 +470,6 @@ def run(self, edit, **args):
-                #         map = generate(source)
-                #     map_syntax = syntax
-
                 else:
```
removed comments

mg979 commented on code_map.py:
```diff
@@ -331,32 +335,6 @@ def navigate_to_line(map_view, give_back_focus=False):
-    w.run_command("set_layout", layout)
-    sublime.set_timeout(lambda: Mapper.block_max_pane(False), 10)            
-
 # =============================================================================
```
this was duplicated

mg979 commented on code_map.py:
```diff
@@ -620,50 +592,25 @@ def run(self, edit, from_view=False):
-                refresh_map_for(self.view)
+                refresh_map_for(self.view, from_view)
                 synch_map(self.view)
 
```
this takes the arg from the command, otherwise it doesn't work

mg979 commented on code_map.py:
```diff
@@ -620,50 +592,25 @@ def run(self, edit, from_view=False):
-            groups = w.num_groups()
-            w.run_command("set_layout", layout)
-            sublime.set_timeout(lambda: Mapper.block_max_pane(False), 10)
-            return (groups + 1)
```
moved on top

mg979 commented on code_map.py:
```diff
@@ -720,10 +667,14 @@ def focus_source_code():
+            # clear temp views variables
+            TEMP_VIDS, TEMP_VIEWS = [], {}
+            CURRENT_TEMP_ID = None
+
```
Temp views are resetted if you close the map

mg979 commented on code_map.py:
```diff
@@ -808,22 +762,14 @@ def on_close(self, view):
+                if settings().get('close_empty_group_on_closing_map', False):
+                    sublime.set_timeout_async(lambda: win(
+                        ).run_command('code_map_ensure_group_closed'))
                 return
```
the on_close listener wasn't launching the command. Sorry if I stress a bit on this, but it causes problems in certain configurations(including mine). It's related to plugins that close empty groups on their own. In pluginless Sublime that works fine, but with certain plugins it can lead to crashes. Running a delayed window command fixes it in my installation, and works fine also for clean installation.

oleg-shilo commented:
Seems OK. I also tested it quickly no new issues. Will wait for you for those 1 and 3(4) form the testing feedback. 

BTW I only mentioned 3(and 4) to you because I thought that it was some how introduced with the latest changes. But I can be wrong. Please let me know if the focus management problems were introduced by some of my old code and I will have a look at it independently.

mg979 commented:
I am not sure about 3 and 4. Maybe you can try to checkout master and see if it happens. I'm trying to add support for your favorites plugin, so that they can stay in the same group. If it works in an uncomplicated way, I'll show it.

oleg-shilo commented:
Not a problem. Just find the compromise for 1 and I will handle the rest. I will not have the opportunity until tomorrow though.
Thank you.

mg979 commented:
Just one thing if things are ok and there are not surfacing bugs.. If you plan to merge it on master sooner or later(maybe after some more testing) maybe you can avoid to merge all commits because they're a lot, I don't know how these things work but I think it would be better if there were a single commit with all the new files, the single commits would stay in the dev branch.

oleg-shilo commented:
Agree. If I cannot figure out how to collapse the commits I will just brutally overwrite master source code with the dev's one and commit with an informative comment. I am sure it's not very nice things to do but... no big deal for me.

oleg-shilo commented:
Cool, it was fun. I've done it. My Git has definitely improved. :)
 
I managed to squash all your PR related commits (along with mine) and now `dev` has the all latest changes. Note I also brought manually the 'code_map.py' from your `just_to_see` branch including my fix for that Focus Management problem (yes it is fixed now).

Thus I suggest you sync and pull all the changes from dev if you still plan to do any work for it. 
I am happy with the codebase and almost ready to push to master and make the release. 

I'll wait a few days just in case if you want to push something last minute.

The only one remaining cosmetic issue is that "view is not bound" error message:
![image](https://user-images.githubusercontent.com/16729806/31212532-cd1f951c-a9ec-11e7-93e5-a8ad4abf0b60.png)

The message is totally OK but the suggestion part of it is not. Trying to refresh the map will fail to generate the map in 100% cases (the view document is just incompatible). Thus suggesting it to the user is not entirely appropriate. 

I can remove the "suggestion" part of the message (in `dev`) however I wanted you to confirm that my above interpretation is correct.

mg979 commented:
> The only one remaining cosmetic issue is that "view is not bound" error message

The message is [there](https://github.com/mg979/sublime-codemap/blob/c571b5e289193fc9634b1eb14ad488b6220151ce/code_map.py#L707), you can change it with a simple "Not a file.". Yes right now only files whose syntax is among the defined syntaxes in the settings can be mapped, not even universal mapper. I think it's better because one could map big/wrong files by mistake.

> I'll wait a few days just in case if you want to push something last minute

The only thing I can think of right now is [this timeout](https://github.com/mg979/sublime-codemap/blob/c571b5e289193fc9634b1eb14ad488b6220151ce/code_map.py#L766), I wonder if it were better placed directly [in the command](https://github.com/mg979/sublime-codemap/blob/c571b5e289193fc9634b1eb14ad488b6220151ce/code_map.py#L727) because wrong stuff in listeners crashes often Sublime. But it didn't give me problems for now, so maybe just to keep in mind if problems arise.

mg979 commented:
There is some problem with the latest changes you made. It makes ST very unstable when switching projects in my installation(plugin host stops working and even crash). Surely it has to do with the listener. Could you please just put in dev only the changes up to this fake PR without your latest changes? If necessary, delete dev branch, create it again from master and squash commits again, but without your changes, there is something that must be fixed.

oleg-shilo commented:
Done. The `dev` now contains your original `just_to_show` copy of _code_map.py_.
> Could you please just put in dev only the changes up to this fake PR without your latest changes?

Note that I only manually brought _code_map.py_ (with my changes) from `just_to_show` to `dev`. Assuming that your changes are already in `dev`. Please let me know if it is the case. May be it is the problem?

Though I want to actually understand and fix the cause of the problem. Surely crashing is worse than uncontrollable active document switching but I'd rather have them both properly fixed. Can you please provide a test-case that demonstrates the problem (if possible)? So I can investigate.

mg979 commented:
Changing  functions in a listener and not restarting ST may be alone a cause for crashes, so maybe this was the only cause, but maybe not. In any case, it's better to have in dev what's sure not to cause issues, sorry I was very alarmed by the crash. If there's some refinement to be made, we start from something that was running smoothly. I hope you agree :)

mg979 commented:
I don't know exactly what happened, I will run some tests. I think the main problem is that when you switch project, the layout/groups/files change very fast, and if the listener tries to close some group or do something on stuff that doesn't exist anymore, it can break something.

oleg-shilo commented:
Yes I agree but I am not sure that the current `dev` is running smoothly. Even though I removed my "Focus Management" changes. 

I checked again  that "Focus Management" change and I  can confirm that it practically did not involve listener. Only a non-aggressive prevention of the _map refresh_ when the source and the map view are in the same group. Thus quite possible the the cause was already in the solution.  

Anyway, I will wait for your report on the `dev` stability and resume gentle reintroduction of "Focus Management" change when the branch is indeed running smoothly.

As for calling  `set_timeout_async` from listener I totally agree that it needs to go to the command.

mg979 committed:
showing n.2

oleg-shilo commented:
Forgot to mention. I manually verified (with KDif3) that the content of `dev` and `just_to_show` is currently identical.

mg979 commented on code_map.py:
```diff
@@ -229,7 +233,11 @@ def generate_from(file):
+    transient = view == w.transient_view_in_group(w.active_group())
+
+    if not map_view or widget or transient or w.is_sidebar_visible():
         return
```
skip also for transient views

mg979 commented on code_map.py:
```diff
@@ -290,6 +298,21 @@ def go():
+    v.show_at_center(line.a)
+
+# -----------------
+
```
moved focus_source_code() here, so no need to repeat. also moved scroll() here because it didn't scroll left when double-clicking on map, only when navigating

mg979 commented on code_map.py:
```diff
@@ -591,12 +609,20 @@ def run(self, edit, from_view=False):
+                    # sync doc -> map
+                    refresh_map_for(self.view, from_view)
+                    synch_map(self.view)
 
```
this is yours #1

mg979 commented on code_map.py:
```diff
@@ -659,10 +678,9 @@ def focus_source_code():
+            # cannot use CodeMapListener.active_view as it will be immediately
+            # overwritten by other views from the new group being activated
+            sublime.set_timeout_async(lambda: w.focus_view(current_view))
 
```
it was true, I wrote it a bit differently because I moved focus_source_code() above

mg979 commented on code_map.py:
```diff
@@ -727,13 +741,17 @@ def run(self, edit):
+                reset_layout(reduce=True)
+
+        sublime.set_timeout_async(delay)
 
```
moved the timeout in the command

mg979 commented on code_map.py:
```diff
@@ -788,12 +798,21 @@ def on_post_save_async(self, view):
+            elif view != CodeMapListener.active_view:
 
                 CodeMapListener.active_view = view
                 refresh_map_for(view)
```
this is yours #2, i made it bit simpler, can you tell me if it works?

mg979 commented:
It was probably a false alarm, sorry. I should have restarted ST. Anyway I think it's better not to add new stuff together with that large commit. I made a new commit here, with your changes plus some little fixes (included timeout in command).

oleg-shilo commented:
> Anyway I think it's better not to add new stuff together with that large commit. 

Amen to that :)

> I made a new commit here, with your changes plus some little fixes (included timeout in command).

Can you please confirm that the branch `just_to_see` is ready and I can try to smoke test it?

mg979 committed:
small change

mg979 commented on code_map.py:
```diff
@@ -798,9 +798,7 @@ def on_post_save_async(self, view):
-        if view == map_view:
+        if ACTIVE and view == get_code_map_view():
             CodeMapListener.map_group = win().get_view_index(view)[0]
 
```
small change here

mg979 commented:
> Amen to that :)

Thanks for understanding :)

I made another small change, now it's ready if you can test it

oleg-shilo commented:
Something went wrong. The map is never generated. Such a dramatic failure is most likely  to my environment conditions. Investigating now...

mg979 commented:
I forgot to add that I disabled map generation when sidebar is open, because it creates maps as you scroll files. Maybe I should remove that. If you close the sidebar map generation works.

oleg-shilo commented:
OK, found the reason. The new condition

oleg-shilo commented:
Yes it is the one :) I tracked it down buy myself

mg979 commented:
Maybe it's not been a good idea. Should it be removed?

oleg-shilo commented:
I think so.

oleg-shilo commented:
BTW, what is this `Window.transient_view_in_group`? I cannot find any documentation on it.

mg979 commented:
Sorry, really a bad idea, it wouldn't work at all with open sidebar.

mg979 commented:
`Window.transient_view_in_group` is the transient view that is created by quick panel. For example "go to anything" command (ctrl+p), or [File History](https://packagecontrol.io/packages/File%20History) plugin. They open a file and CodeMap maps it without that condition. I was looking for a way to prevent that behaviour, I found this command in another plugin.

mg979 commented:
By the way, are your changes working with this version?

oleg-shilo commented:
Seems like a good improvement. 

Yes my changes work but with one exception. It still changes the active doc if the toggling (e.g. closing map view) was triggered while another doc from the map group has focus.   

I feel like it is something that I have fixed ("Focus Management"). Thus if you feel that it is unlikely that it's introduced with your latest commits, I can redo it. But of course, only when you signal it's OK for me to step in :)

mg979 commented:
I changed a bit on_activated_async(), maybe this is the cause. I don't think I want to make more changes, from my part, for now, I can't think of anything else.

oleg-shilo commented:
OK then. 
It is already reasonable stable and addresses ~70% of that _FocusManagement_ issues. The rest is a very infrequent use-cases.  Let's seal it for now. Will keep using/testing for a couple of days and if no other issued detected then we will release it.

In background I will investigate that 30% of problems but will introduce the fix (if found) only after that graceful period.

Do you want me to squash your  commits?

oleg-shilo commented:
Oh, sorry I cannot. Only you can do. It's not my branch.

mg979 commented:
Maybe I understood what happens. I cannot test it because I don't know how to reproduce it but:

when you close through command (alt+m, alt+m) this is run and updates current view for focus_source_code():
`CodeMapListener.active_view = view`

If you close the map with the mouse this doesn't run, so `CodeMapListener.active_view` is still the old view.

mg979 commented:
My commits are done for now. How is it that I can squash them? can't you just take the file and overwrite, then squash them?

oleg-shilo commented:
Txs. Have a nice day.

mg979 commented:
Ok nice day to you too.

mg979 commented:
By the way I'm liking your Favorites plugin, I thought you could make this:

```
class show_favorites(sublime_plugin.TextCommand):
    def run(self, edit, focus=False):

        w = sublime.active_window()
        fav_view = get_panel_view()
        if focus and fav_view:
            w.focus_view(fav_view)
            w.focus_view(self.view)
        else:
            show_panel.run(self, edit)
```

Then a keybinding:

`    { "keys": [...], "command": "show_favorites", "args": {"focus": true} },`

So that you can put map and favorites in the same group, then synch map to focus on it, and  this command to focus on favorites.

Also this in class show_panel:

```
        if not panel_view:

            panel_group = -1
            last_group = window.num_groups()-1
            # look for CodeMap in last group
            for view in window.views_in_group(last_group):
                if os.path.basename(view.file_name()) == 'Code - Map':
                    panel_group = last_group

            # CodeMap not found
            if panel_group == -1:
                show_in_new_group = settings().get("show_in_new_group", True)

                if not show_in_new_group:
                    if groups == 1:
                        set_layout_columns(2)
                        groups = window.num_groups()

                else:
                    panel_group = create_panel_group()
```

If then you have keybindings like:

`    { "keys": ["alt+f", "alt+f"], "command": "show_favorites", "args": {"focus": true} },`

This opens Favorites in the same CodeMap group, but if you press it again it just focuses on it. Old command to close if you need (or use command palette):

`    { "keys": ["alt+f", "alt+x"], "command": "show_favorites"`

oleg-shilo commented:
That last 30%. Solved. Was a simple case of missing `alone_in_group` test:

```Python
def reset_layout(reduce):
    ...
    map_view = get_code_map_view()
    alone_in_group = len(w.views_in_group(get_group(map_view))) == 1

    if map_view and alone_in_group:
        w.set_view_index(map_view, 0, 0)
    ...
```
Though I will wait with it until you create the PR so I can merge your `just_to_see` into my `dev`.

mg979 commented:
There's still a problem before I can make a PR. It's not very stable yet when I switch projects. I'll try to find a fix because it's a very big problem, I don't want crashes. I also have them with the commit that I considered stable, so I have to find the cause.

oleg-shilo commented:
Txs, I will review Favorites when we are done with this one. 
BTW, with your future PR, there is no pressure at all. So take your time.

I also discovered another nasty artifact. If node, just below the currently selected one, is double-clicked it does not trigger the new selection reliably: 
![code_map_3](https://user-images.githubusercontent.com/16729806/31233952-6d014638-aa39-11e7-9f61-157e9ce9c509.gif)
Though if it is not the next one but the further below node then it's all good.

When I reflect at all this experience I cannot stop being disappointing with ST3  almost religious stubbornness with resisting true panels. :(
I like so much more VSCode for that.

mg979 commented:
> Though if it is not the next one but the further below node then it's all good.

Mmm it's actually happening with all of them, it only depends on how fast you click another line. But not really always. It's probably because there are some async timeouts that need their time, maybe there are timeouts in sequence that also control focus. I don't think it's a big issue. I remember I set a timeout because it wasn't selecting the line anymore, just a word. I see if I can improve this too.

oleg-shilo commented:
I just fixed it. Will share in 5 mins. 
It is one of those actions from the listeners cases. Separation to the command fixed it.

oleg-shilo commented:
```Python
class navigate_to_line_async(sublime_plugin.WindowCommand):

    def run(self):

        def do():
            map_view = get_code_map_view()
            navigate_to_line(map_view , give_back_focus = not CodeMapListener.navigating)

        sublime.set_timeout_async(do, 10)

# =================================

def on_text_command(self, view, command_name, args):
    '''Process double-click on code map view'''

    double_click = (command_name == 'drag_select' and 'by' in args.keys() and args['by'] == 'words')

    if double_click:
        if ACTIVE and view.file_name() == code_map_file():
            win().run_command("navigate_to_line_async")
            return ("code_map_select_line", None)
```

mg979 commented:
It makes it a bit slower but it works.. Though I'm not sure it's better... The other problem is much bigger, though, still trying to find a solution, maybe almost there.

oleg-shilo commented:
In my environment I didn't notice slowing down and it worked always. But I do feel like it is a work around but not a fix for the cause.

Anyway, as a side exercise I made this little _marshaler_ that marshals a given routine into the sublime command and executes it asynchronously. It can be used instead of `navigate_to_line_async` and for any other heavy routines invoked from the listener:
 
```python
# sample:
# marshaler.invoke(lambda: print('test'))

import uuid

class marshaler(sublime_plugin.WindowCommand):
    _actions = {}

    def invoke(action, delay=10):
        action_id = str(uuid.uuid4())
        marshaler._actions[action_id] = (action, delay)
        win().run_command("marshaler", {"action_id": action_id} )

    def run(self, **args):
        action_id = args['action_id']
        action, delay = marshaler._actions.pop(action_id)
        sublime.set_timeout_async(action, delay)
```

mg979 commented:
I'm almost done with my changes. I think I fixed that problem, at least I hope. A couple of things have changed(not dramatically). While I was there, I fixed a bug with navigation (error when starting from a region before the first map node) that was there since the beginning and I could finally eradicate. I included your fix from https://github.com/oleg-shilo/sublime-codemap/pull/14#issuecomment-334475502, but not that async command. I'll leave it to you if you want to incude it. I will make a commit, if for you it's ok I'll make a PR.

mg979 committed:
fixes(stability, navigation) and some changes in behaviour

mg979 commented on code_map.py:
```diff
@@ -33,7 +33,6 @@
     global MAPPERS, ACTIVE, using_universal_mapper
 
-    ACTIVE = True if get_code_map_view() else False
     using_universal_mapper = True
```
(Changed in last commit, CodeMap starts normally, you can ignore this)

Now CodeMap starts in inactive state, this means that if you restart ST and CodeMap panel is open, it will be inactive, you must synch. Synch command is enough, not needed to close/reopen. This change has a benefit, that CodeMap doesn't start mapping files when ST is restarted and loads all files, but the drawback that it must be synched manually after start. If you don't like it, I think it's safe enough to restore old behaviour, in this point(but not when switching projects).

mg979 commented on code_map.py:
```diff
@@ -181,7 +189,12 @@ def reset_layout(reduce):
+        alone_in_group = len(w.views_in_group(get_group(map_view))) == 1
+    except:
+        alone_in_group = True
+
```
I added a try block because I had an error in an occasion. map_view had been closed and was not there. I think this could need some rework, maybe just check if the last group has no more views.

If it breaks your change, you can try:

`    alone_in_group = len(w.views_in_group(CodeMapListener.map_group)) == 1`

instead of the try block

mg979 commented on code_map.py:
```diff
@@ -249,8 +262,8 @@ def generate_from(file):
         clear_map_selection()
-    elif not file:
-        win().run_command('code_map_temp_view_msg')
+    # elif not file:
```
I removed totally the message for non-files, it was causing instability

mg979 commented on code_map.py:
```diff
@@ -355,6 +367,14 @@ def navigate_to_line(map_view, give_back_focus=False):
+            map_view.sel().add(Region(0, 0))
+            Nav.highlight_line(map_view)
+
         if not give_back_focus:
```
this solves the navigation bug

mg979 commented on code_map.py:
```diff
@@ -570,6 +590,14 @@ class navigate_code_map(sublime_plugin.TextCommand):
+            else:
+                win().status_message('  CodeMap is curently closed.')
+            return
+
```
A warning if CodeMap is not active, instead of an error

mg979 commented on code_map.py:
```diff
@@ -748,6 +778,7 @@ def delay():
 
         sublime.set_timeout_async(delay)
+        sublime.set_timeout_async(focus_source_code)
 
```
added this to focus back when closing map with mouse, I didn't have problems for now

mg979 commented on code_map.py:
```diff
@@ -759,12 +790,13 @@ class CodeMapListener(sublime_plugin.EventListener):
+                CodeMapListener.skip = True
+            elif CodeMapListener.skip:
+                CodeMapListener.skip = False
 
```
I added ACTIVE checks for all listeners

mg979 commented on code_map.py:
```diff
@@ -828,12 +860,34 @@ def on_text_command(self, view, command_name, args):
+            # if command_name == "project_manager":
+            #     window.run_command('show_code_map')
+        return None
 
```
(in the last commit CodeMap is only paused, not stopped)

If project is switched, CodeMap is inactivated(but not closed). When a new project is loaded, if the panel is open it must be synched first.

mg979 commented on code_map.py:
```diff
@@ -601,10 +629,16 @@ def run(self, edit, direction=None, start=False, stop=False, fast=False):
+
+        if not ACTIVE and get_code_map_view():
+            ACTIVE = True
 
```
(not very important in the last commit)

I modified the synch map command, so that it activates the map if the panel is open but the map is inactive.

mg979 committed:
a better fix for stability that doesn't change old behaviour

mg979 commented on code_map.py:
```diff
@@ -801,10 +802,15 @@ def on_deactivated(self, view):
+            PAUSED = False
+            ACTIVE = True
+
     # -----------------
```
When new files start loading, CodeMap is activated again

mg979 commented on code_map.py:
```diff
@@ -866,7 +872,7 @@ def on_text_command(self, view, command_name, args):
+        global TEMP_VIDS, TEMP_VIEWS, CURRENT_TEMP_ID, ACTIVE, PAUSED
 
         reset = ["prompt_open_project_or_workspace",
                  "prompt_switch_project_or_workspace",
```
when switching project CodeMap is paused until new files are loaded, so its listeners don't work

mg979 committed:
removed PAUSED state, made it simpler, no crashes, behaviour as before, one fix for map group, for me it's ready
